src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleMetadata.cs

```diff
@@ -84,5 +107,31 @@ public static partial class JsonSerializer
 
             return writtenMetadataName;
         }
+
+        /// <summary>
+        /// Used by polymorphic converters that are handling references for values that are boxed structs.
+        /// </summary>
+        internal static bool TryWriteReferenceForBoxedStruct(object currentValue, ref WriteStack state, Utf8JsonWriter writer)
+        {
+            Debug.Assert(currentValue.GetType().IsValueType);
+
+            string referenceId = state.ReferenceResolver.GetReference(currentValue, out bool alreadyExists);
+            Debug.Assert(referenceId != null);
+
+            if (alreadyExists)
+            {
+                writer.WriteStartObject();
+                writer.WriteString(s_metadataRef, referenceId);
+                writer.WriteEndObject();
+            }
+            else
+            {
+                // Since we cannot run `ReferenceResolver.GetReference` twice for newly encountered instances,
```

Comment:
Why can't we write `referenceId` here like we do for other cases (the boxed value's id). Adding the `BosedStructReferenceId` doesn't seem right -- would it work for nested object \ collection cases since it is not stack-based (i.e. doesn't know about nesting like `Order.Customer.Address.Line1`).

Reply:
> Why can't we write referenceId here like we do for other cases (the boxed value's id).

Because the way `referenceId` is written on the wire gets dictated by the converter that ends up handling that particular object: object converters simply prepend it to the list of properties whereas collection converters need to nest their payload under a `$values` property.

> Adding the BoxedStructReferenceId doesn't seem right -- would it work for nested object \ collection cases since it is not stack-based (i.e. doesn't know about nesting like Order.Customer.Address.Line1).

Since this corner case only manifests itself in the context of polymorphic converters, it is given that the next converter will be consuming that field before any child nodes get serialized. I could have placed this property in the stackframe struct instead, but that would have grown each frame by one word. I'll push a commit that adds a `Debug.Assert(state.BoxedStructReferenceId != null);` check before this line to make sure that invariant keeps being honored.

Reply:
> it is given that the next converter will be consuming that field before any child nodes get serialized

Does that mean when we have a `System.Object` property and the underlying type is a POCO or collection that is read as a single value and the Stream-based "continuation" mode doesn't apply? Do we have tests that verify this (large collections\objects + system.object polymorphism + reference handling + Stream"?

Reply:
> Does that mean when we have a System.Object property and the underlying type is a POCO or collection that is read as a single value and the Stream-based "continuation" mode doesn't apply?

No, async serialization still works under this model. The reason this invariant holds (for the case of serialization) is because metadata is written [at the start and never as part of a continuation](https://github.com/dotnet/runtime/blob/aa4290f3f0aadf86882a439d124db3215851760b/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableDefaultConverter.cs#L242).

> Do we have tests that verify this (large collections\objects + system.object polymorphism + reference handling + Stream"?

Our tests do catch that particular case, however as a general note I would say that we could do better w.r.t. async test coverage. I did catch a few bugs related to polymorphism+async when working on the prototype. I'm planning on adding testing infrastructure to better stress test our async code during the RC1 timeframe.